src/hotspot/share/cds/heapShared.cpp

```diff
@@ -1401,12 +1405,17 @@ class HeapShared::ReferentPusher: public BasicOopIterateClosure {
   template <class T> void do_oop_work(T *p) {
     oop obj = RawAccess<>::oop_load(p);
     if (!CompressedOops::is_null(obj)) {
-      size_t field_delta = pointer_delta(p, _referencing_obj, sizeof(char));
+      int field_offset = pointer_delta_as_int((char*)p, cast_from_oop<char*>(_referencing_obj));
```

Comment:
The RawAccess oop load above should probably be a HeapAccess<ON_UNKNOWN_OOP_REF>::oop_load_at. By using the unknown oop ref and supplying the field offset, we allow GCs to infer the actual reference strength in the backend when loading and apply appropriate barriers for non-strong references, while also applying appropriate load barriers for conc GCs.